lib/image_composition.dart

```diff
@@ -0,0 +1,167 @@
+import 'dart:async';
+import 'dart:ui';
+
+import 'extensions.dart';
+
+export 'extensions.dart';
+
+class _Composed {
+  /// The image that will be composed.
+  final Image image;
+
+  /// The position where the [image] will be composed.
+  final Vector2 position;
+
+  /// The source on the [image] that will be composed.
+  final Rect source;
+
+  final double rotation;
+
+  final Vector2 anchor;
+
+  final bool isAntiAlias;
+
+  /// The [BlendMode] that will be used when composing the [image].
+  final BlendMode blendMode;
+
+  _Composed(
+    this.image,
+    this.position,
+    this.source,
+    this.rotation,
+    this.anchor,
+    this.isAntiAlias,
+    this.blendMode,
+  );
+}
+
+class ImageComposition {
+  /// The values that will be used to compose the image
+  final List<_Composed> _composes = [];
+
+  /// The [defaultBlendMode] can be used to change how each image will be
+  /// blended onto the composition. Defaults to [BlendMode.srcOver].
+  final BlendMode defaultBlendMode;
+
+  /// The [defaultAntiAlias] can be used to if each image will be anti aliased.
+  final bool defaultAntiAlias;
+
+  ImageComposition({
+    this.defaultBlendMode = BlendMode.srcOver,
+    this.defaultAntiAlias = false,
+  })  : assert(defaultBlendMode != null, 'defaultBlendMode can not be null'),
+        assert(defaultAntiAlias != null, 'defaultAntiAlias can not be null');
+
+  /// Add an image to the [ImageComposition].
+  ///
+  /// The [image] will be added at the given [position] on the composition.
+  ///
+  /// An optional [source] can be used to only add the data that is within the
+  /// [source] of the [image].
+  ///
+  /// An optional [rotation] (in radians) can be used to rotate the image when it
+  /// gets added to the composition. You can use [anchor] to set the point on
+  /// which it will rotate.
+  ///
+  /// By default the [anchor] will be the [source.width] and [source.height]
+  /// divided by `2`.
+  ///
+  /// [isAntiAlias] can be used to if the [image] will be anti aliased. Defaults
+  /// to [defaultAntiAlias].
+  ///
+  /// The [blendMode] can be used to change how the [image] will be blended onto
+  /// the composition. Defaults to [defaultBlendMode].
+  void add(
+    Image image,
+    Vector2 position, {
+    Rect source,
+    double rotation = 0,
+    Vector2 anchor,
+    bool isAntiAlias,
+    BlendMode blendMode,
+  }) {
+    assert(image != null, 'Image is required to add to the Atlas');
+    assert(position != null, 'Position is required');
+    assert(
+      source == null ||
+          source.width <= image.width &&
+              source.height <= image.height &&
+              source.top + source.height <= image.height &&
+              source.top >= 0 &&
+              source.left + source.width <= image.width &&
+              source.left >= 0,
```

Comment:
I wonder if there is an easy to to convert image => Rect so you can just use an existing Rect method to check this

Reply:
Maybe an Image extension?

Reply:
Agreed might be useful to add, and would make the code more readable(and fool proof)

Reply:
I have a few other extension methods for Images in my personal project that would be beneficial for image manipulation. For instance changing the darkness or luminate an image.

Maybe that could be a PR on its own?

Reply:
As discussed yesterday, I will follow up on this with a separate PR after this has been merged that will introduce an image extension class